k8sapiserver/k8sapiserver.go

```diff
@@ -81,199 +71,165 @@ func StartAPIServer(kubeAPIServerURL, etcdURL, frontendURL string) (*restclient.
 		s.Close()
 		klog.Info("destroyed API server")
 	}
-	return cfg, shutdownFunc, nil
-}
 
-func defaultOpenAPIConfig() *openapicommon.Config {
-	openAPIConfig := genericapiserver.DefaultOpenAPIConfig(generated.GetOpenAPIDefinitions, openapi.NewDefinitionNamer(legacyscheme.Scheme))
-	openAPIConfig.Info = &spec.Info{
-		InfoProps: spec.InfoProps{
-			Title:   "Kubernetes",
-			Version: "unversioned",
-		},
-	}
-	openAPIConfig.DefaultResponse = &spec.Response{
-		ResponseProps: spec.ResponseProps{
-			Description: "Default Response.",
-		},
+	cfg := &restclient.Config{
+		Host:          s.URL,
+		ContentConfig: restclient.ContentConfig{GroupVersion: &schema.GroupVersion{Group: "", Version: "v1"}},
+		QPS:           5000.0,
+		Burst:         5000,
 	}
-	openAPIConfig.GetDefinitions = generated.GetOpenAPIDefinitions
 
-	return openAPIConfig
+	return cfg, shutdownFunc, nil
 }
 
-//nolint:funlen
-func NewControlPlaneConfigWithOptions(serverURL, etcdURL, frontendURL string) *controlplane.Config {
-	etcdOptions := options.NewEtcdOptions(storagebackend.NewDefaultConfig(uuid.New().String(), nil))
-	etcdOptions.StorageConfig.Transport.ServerList = []string{etcdURL}
-
-	storageConfig := kubeapiserver.NewStorageFactoryConfig()
-	storageConfig.APIResourceConfig = serverstorage.NewResourceConfig()
-	completedStorageConfig, err := storageConfig.Complete(etcdOptions)
-	if err != nil {
-		panic(err)
+func setUpHandlerAndRun(aggregatorServer *apiserver.APIAggregator, s *httptest.Server, h *APIServerHolder) func() {
+	var m *controlplane.Instance
+	stopCh := make(chan struct{})
+	closeFn := func() {
+		if m != nil {
+			if err := m.GenericAPIServer.RunPreShutdownHooks(); err != nil {
+				klog.Errorf("failed to run pre-shutdown hooks for api server: %v", err)
+			}
+		}
+		close(stopCh)
+		s.Close()
 	}
-	storageFactory, err := completedStorageConfig.New()
-	if err != nil {
-		panic(err)
+	m = &controlplane.Instance{
+		GenericAPIServer: aggregatorServer.GenericAPIServer,
 	}
+	h.SetAPIServer(m)
 
-	genericConfig := genericapiserver.NewConfig(legacyscheme.Codecs)
-	kubeVersion := version.Get()
-	if len(kubeVersion.Major) == 0 {
-		kubeVersion.Major = "1"
-	}
-	if len(kubeVersion.Minor) == 0 {
-		kubeVersion.Minor = "22"
-	}
-	genericConfig.Version = &kubeVersion
+	errCh := make(chan error)
```

Comment:
Is this `errCh` needed?

Reply:
It is not needed here, I have updated the PR with the errch removed. However I am not sure whether we should catch any err return from prepared.run()

Reply:
I think we should log errors at least. Could you add logging for error here?